<a href="https://colab.research.google.com/github/Lalo1005/eduardo-git/blob/master/Python_Data_Science_Numpy_inicial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color=green> PYTHON PARA DATA SCIENCE - NUMPY
---

# <font color=green> 1. INTRODUCCIÓN A PYTHON
---

# 1.1 Introducción

> Python es un lenguaje de programación de alto nivel con soporte a múltiples paradigmas de programación. Es un proyecto *open source* y desde su surgimiento, en 1991, viene convirtiéndose en uno de los lenguajes de programación interpretados más populares.
>
> En los últimos años Python desarrolló una comunidad activa de procesamiento científico y análisis de datos y viene destacándose como uno de los lenguajes más relevantes cuando el asunto es ciencia de datos y machine learning, tanto en el entorno acadêmico como también en el entorno laboral.

# 1.2 Instalación y entorno de trabajo

### Instalación Local

### https://www.python.org/downloads/
### o
### https://www.anaconda.com/distribution/

### Google Colaboratory

### https://colab.research.google.com

### Verificando versión

In [ ]:
!python -V

Python 3.10.12


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


**Numpy es una biblioteca para cálculo numérico.**

# 1.3 Trabajando con Numpy

In [ ]:
import numpy as np

**Para cargar el arhivo carros-km.txt usamos el método .nploadtxt()**

Para que numpy sepa el tipo de archivo agregamos dtype.

In [ ]:
km = np.loadtxt('/content/drive/MyDrive/data/carros-km.txt', dtype=int)

In [ ]:
km

array([ 44410,   5712,  37123,      0,  25757,  10728,      0,  77599,
        99197,  37978,  12859,   8052,  89773,      0,  41457, 115607,
        46449,      0,  37086,  15173, 101193,      0,  98079, 102959,
            0,      0,   5795,      0,  58848,  94381,  30163,  53332,
        17720,  33808,  90684,  43975,      0,      0,   5526,      0,
        93415,  40762,      0,  86302,      0,   9755,  69945,   2395,
            0,  80349,  85554,  50496,  67716,  93947,  35345,  81007,
       119513,      0,      0,      0,      0,      0, 118895,  48509,
       100912,  95649,      0,  90495,      0,  29132,  23802,  84992,
        54395,  26731,  44329, 118236, 113808,    610,      0,      0,
        12887,  79607,  90924,  42733,      0,      0, 117714, 113885,
            0,  30511,  74867, 119760,   8356,  64247,  88661,   4539,
       110116,  33215,  92001,      0,  81708,  70641,      0,  91277,
        26544,  52596,  47503,  89056,  28834, 110564,  56638,  17357,
      

In [ ]:
años = np.loadtxt('/content/drive/MyDrive/data/carros-años.txt', dtype=int)

In [ ]:
años

array([2003, 1991, 1990, 2020, 2006, 2012, 2020, 2009, 2010, 2011, 2002,
       2007, 2001, 2020, 2009, 2016, 2012, 2020, 2016, 2017, 2002, 2020,
       1994, 1991, 2020, 2020, 2016, 2020, 2002, 1991, 1997, 1995, 2018,
       1991, 2014, 1995, 2020, 2020, 2009, 2020, 2013, 2003, 2020, 2002,
       2020, 2002, 2004, 2001, 2020, 2003, 2009, 2006, 2000, 2006, 2010,
       1994, 2007, 2020, 2020, 2020, 2020, 2020, 2017, 2011, 2001, 1996,
       2020, 1994, 2020, 2007, 1994, 2007, 1997, 2006, 2013, 1994, 2008,
       2008, 2020, 2020, 1999, 2007, 1993, 1995, 2020, 2020, 2016, 2011,
       2020, 2002, 2013, 2017, 2016, 2008, 2018, 2003, 2003, 2004, 2010,
       2020, 2006, 2001, 2020, 2015, 2018, 2017, 1992, 2001, 1993, 1995,
       2003, 2014, 2008, 2013, 2020, 2017, 2020, 2020, 2006, 2015, 2007,
       2005, 2017, 1998, 2020, 1999, 2020, 2009, 2003, 1994, 2020, 1994,
       2020, 2015, 1996, 1997, 1993, 2020, 1992, 1995, 2020, 1990, 2020,
       2020, 1991, 1990, 2012, 2007, 2013, 2000, 20

# $$km_{promedio} = \frac {km_{total}}{(Año_{actual} - Año_{fabricación})}$$

### Obtener el promedio de kilometros recorridos por año

In [ ]:
km_promedio = km / (2023 - años)

In [ ]:
km_promedio

array([2.22050000e+03, 1.78500000e+02, 1.12493939e+03, 0.00000000e+00,
       1.51511765e+03, 9.75272727e+02, 0.00000000e+00, 5.54278571e+03,
       7.63053846e+03, 3.16483333e+03, 6.12333333e+02, 5.03250000e+02,
       4.08059091e+03, 0.00000000e+00, 2.96121429e+03, 1.65152857e+04,
       4.22263636e+03, 0.00000000e+00, 5.29800000e+03, 2.52883333e+03,
       4.81871429e+03, 0.00000000e+00, 3.38203448e+03, 3.21746875e+03,
       0.00000000e+00, 0.00000000e+00, 8.27857143e+02, 0.00000000e+00,
       2.80228571e+03, 2.94940625e+03, 1.16011538e+03, 1.90471429e+03,
       3.54400000e+03, 1.05650000e+03, 1.00760000e+04, 1.57053571e+03,
       0.00000000e+00, 0.00000000e+00, 3.94714286e+02, 0.00000000e+00,
       9.34150000e+03, 2.03810000e+03, 0.00000000e+00, 4.10961905e+03,
       0.00000000e+00, 4.64523810e+02, 3.68131579e+03, 1.08863636e+02,
       0.00000000e+00, 4.01745000e+03, 6.11100000e+03, 2.97035294e+03,
       2.94417391e+03, 5.52629412e+03, 2.71884615e+03, 2.79334483e+03,
      

In [ ]:
type(km_promedio)

numpy.ndarray

# <font color=green> 2. CARACTERÍSTICAS BÁSICAS DEL LENGUAJE
---

# 2.1 Operaciones matemáticas

### Operadores aritméticos: $+$, $-$, $*$, $/$, $**$, $\%$, $//$

### Suma ($+$)

In [ ]:
2+2

4

### Resta ($-$)

In [ ]:
2-2

0

### Multiplicación ($*$)

In [ ]:
2*3

6

### División ($/$) y ($//$)
La operación división siempre devuelve un número con punto decimal

In [ ]:
10/3

3.3333333333333335

In [ ]:
10//3

3

### Exponenciación ($**$)

In [ ]:
2**3

8

### Resto de la división ($\%$)

In [ ]:
10%3

1

Cuando la división no es exacta el resto de la división es el entero faltante para completar el número divido. Ejemplo 10 entre 3 es igual a 3, entonces 3 * 3= 9, 9 + el resto de la división es igual a 10.

In [ ]:
10 % 2

0

Cuando la división es exacta el resto de la división es cero.

### Expresiones matemáticas

In [ ]:
5 * 2 + 3 * 2

16

In [ ]:
(5 * 2) + (3 * 2)

16

In [ ]:
5 * (2 + 3) * 2

50

Primero ejecuto la operación entre parétesis y luego la primera multiplicación y después la segunda multiplicación.

### La variable _

En el modo interactivo, el último resultado impreso en pantalla es atribuído a la variable _

In [ ]:
_

99

In [ ]:
5 * 2

10

In [ ]:
_

99

Adicionando otra operación utilizando la variable _

In [ ]:
_ + 3 * 2

105

Dividiendo en otra operación utilizando la variable _

In [ ]:
_ / 2

49.5

# 2.2 Variables

### Nombres de variables

- Nombres de variables pueden comenzar con letras (a - z, A - Z) o con el símbolo *underscore* (_):

    > Altura
    >
    > _peso
    
- El resto del nombre puede contener letras, números o el símbolo "_":

    > nombre_de_variable
    >
    > _valor
    >
    > dia_28_11_
    

- Los nombres son *case sensitive*:

    > Nombre_De_Variable $\ne$ nombre_de_variable $\ne$ NOMBRE_DE_VARIABLE
    
### <font color=red>Observaciones:
- Existen algumas palabras reservadas del lenguaje que no pueden ser utilizadas como nombres de variables:

| |Lista de palabras <br>reservadas en Python| |
|:-------------:|:------------:|:-------------:|
| and           | as           | not           |
| assert        | finally      | or            |
| break         | for          | pass          |
| class         | from         | nonlocal      |
| continue      | global       | raise         |
| def           | if           | return        |
| del           | import       | try           |
| elif          | in           | while         |
| else          | is           | with          |
| except        | lambda       | yield         |
| False         | True         | None          |

### Declaración de variables

### Operadores de asignación: $=$, $+=$, $-=$, $*=$, $/=$, $**=$, $\%=$, $//=$

Está es la forma como se declara una variable

In [ ]:
año_actual = 2023
año_fabricacion = 2003
km_total = 44410.0

In [ ]:
año_actual

2023

In [ ]:
año_fabricacion

2003

In [ ]:
km_total

44410.0

# $$km_{promedio} = \frac {km_{total}}{(Año_{actual} - Año_{fabricación})}$$

### Operaciones con variables

In [ ]:
km_total/(año_actual - año_fabricacion)

2220.5

Aquí en la variable km_promedio almacenamos el resultado de la ecuación.

In [ ]:
km_promedio = km_total/(año_actual - año_fabricacion)
km_promedio

2220.5

In [ ]:
año_actual = 2023
año_fabricacion = 2003
km_total = 44410.0
km_promedio = km_total/(año_actual - año_fabricacion)
km_promedio

2220.5

In [ ]:
año_actual = 2023
año_fabricacion = 2003
km_total = 44410.0
km_promedio = km_total/(año_actual - año_fabricacion)
# El kilometraje del recorrido es hata el año 2023 igual a 44410.
# Quiero obtener el kilometraje hasta el 2024 entonces le sumo el promedio
km_total = km_total + km_promedio
km_total

46630.5

Existe otra manera de definirlo más fácil

In [ ]:
año_actual = 2023
año_fabricacion = 2003
km_total = 44410.0
km_promedio = km_total/(año_actual - año_fabricacion)

km_total += km_promedio
km_total

46630.5

### Conclusión:
```
"valor = valor + 1" es equivalente a "valor += 1"
```

### Declaración múltiple

In [ ]:
año_actual, año_fabricacion, km_total = 2023,2003, 44410.0

In [ ]:
año_actual

2023

In [ ]:
año_fabricacion

2003

In [ ]:
km_total

44410.0

# 2.3 Tipos de datos

Los tipos de datos especifican como números y caracteres estarán guardados y manipulados dentro de un programa. Los tipos de datos básicos en Python son:

1. **Números**
    1. ***int*** - Enteros
    - ***float*** - Decimales
- **Booleanos** - Contiene valores True o False. Esencial cuando comenzamos a trabajar con declaraciones condicionales
- ***Strings*** - Secuencia de uno o más caracteres que puede incluir letras, números y otros tipos de caracteres. Representa un texto.
- **None**,  **Not a Number - NaN** .- Representa la ausencia de valor

### Números

In [ ]:
año_actual = 2023

In [ ]:
type(año_actual)

int

In [ ]:
km_total = 44410.0

In [ ]:
type(km_total)

float

### Booleanos

In [ ]:
zero_km = True

In [ ]:
type(zero_km)

bool

In [ ]:
zero_km = False

In [ ]:
type(zero_km)

bool

### Strings

In [ ]:
nombre = 'Jetta Variant'
nombre

'Jetta Variant'

In [ ]:
nombre = "Jetta Variant"
nombre

'Jetta Variant'

In [ ]:
nombre = 'Jetta "Variant"'
nombre

'Jetta "Variant"'

In [ ]:
nombre = "Jetta 'Variant'"
nombre

"Jetta 'Variant'"

Para crear texto con mútiples líneas

In [ ]:
carro = '''
  Nombre
  Edad
  Nota
'''

\n significa salto de línea porque no están en una misma línea, están en múltiples líneas.

In [ ]:
carro

'\n  Nombre\n  Edad\n  Nota\n'

In [ ]:
type(carro)

str

### None

Con el comando None indicamos que el valor es nulo, que no tiene valor. No debe mostrar nada.

In [ ]:
kilometraje = None

In [ ]:
kilometraje

In [ ]:
type(kilometraje)

NoneType

# 2.4 Conversión de tipos

In [ ]:
a = 10
b = 20
c = 'Python es '
d = 'genial'

In [ ]:
type(a)

int

In [ ]:
type(b)

int

In [ ]:
type(c)

str

In [ ]:
type(d)

str

In [ ]:
a + b

30

Las variables string no se suman se concatenan.

In [ ]:
c + d

'Python es genial'

No se pueden mezclar los string con los números, porque da error.

In [ ]:
#c + a

* La función int() convierte todo a entero, la función float() convierte todo a decimal y la función string() todo en strings.

### Conversiones de tipo

Funciones int(), float(), str()

Convirtiendo a texto un número

In [ ]:
str(a)

'10'

In [ ]:
type(str(a))

str

In [ ]:
c + str(a)

'Python es 10'

In [ ]:
float(a)

10.0

In [ ]:
var = 3.141592

In [ ]:
int(var)

3

In [ ]:
var = 3.99

In [ ]:
int(var)

3

# 2.5 Indentación, comentarios y formatación de *strings*

### Indentación (Sangría)

En el lenguaje Python los programas son estructurados por medio de indentación. En cualquier lenguaje de programación la práctica de colocar indentación es bastante útil, facilitando la lectura y también el mantenimiento del código. En Python la indentación no es solamente una cuestión de organización y estilo, sino que es un requisito del lenguaje.

In [ ]:
año_actual = 2023
año_fabricacion = 2023

if (año_actual == año_fabricacion):
    print('Verdadero')
else:
    print('Falso')

Verdadero


### Comentarios

Comentarios son extremamente importantes en un programa. Consiste en un texto que describe lo que el programa o una parte específica del programa está haciendo. Los comentarios son ignorados por el interpretador Python.

Podemos tener comentarios de una única linea o de múltiples líneas.

In [ ]:
# Esto es un comentario
año_actual = 2023
año_actual

2023

In [ ]:
# Esto
# es un
# comentario
año_actual = 2023
año_actual

2023

In [ ]:
'''Esto es un
comentario'''
año_actual = 2023
año_actual

2023

In [ ]:
# Definiendo variables
año_actual = 2023
año_fabricacion = 2023

'''
Estructura condicional que vamos
a aprender la próxima clase
'''
if (año_actual == año_fabricacion):   # Probando si la condición es verdadera
    print('Verdadero')
else:                               # Probando si la condición es falsa
    print('Falso')

Verdadero


### Formatación de *strings*

**Técnica muy importante bastante utilizada al mostrar los resultados de nuestros algoritmos.**

## *str.format()*

https://docs.python.org/3.6/library/stdtypes.html#str.format

In [ ]:
print('Hola,  {}')

Hola,  {}


In [ ]:
print('Hola,  {}'.format('Alejandro'))

Hola,  Alejandro


In [ ]:
print('Hola,  {}! Esta es su vista número {} '.format('Alejandro', 32))

Hola,  Alejandro! Esta es su vista número 32 


In [ ]:
print('Hola,  {}! Esta es su vista número {} '.format(32, 'Alejandro'))

Hola,  32! Esta es su vista número Alejandro 


Entonce para mejorar esto declaramos variables dentro del formato format().

In [ ]:
print('Hola,  {nombre}! Esta es su vista número {visita} '.format(visita=32, nombre='Alejandro'))

Hola,  Alejandro! Esta es su vista número 32 


## *f-Strings*

https://docs.python.org/3.6/reference/lexical_analysis.html#f-strings

In [ ]:
nombre = 'Eduardo Martín'
visita = 10

In [ ]:
print('Hola,  {nombre}! Esta es su vista número {visita} ')

Hola,  {nombre}! Esta es su vista número {visita} 


In [ ]:
print(f'Hola,  {nombre}! Esta es su vista número {visita} ')

Hola,  Eduardo Martín! Esta es su vista número 10 


# <font color=green> 3. TRABAJANDO CON LISTAS
---

# 3.1 Creando listas

Listas son secuencias **mutables** que son utilizadas para guardar colecciones de objetos, generalmente homogéneos. Pueden ser construídas de varias formas:
```
- Utilizando un par de corchetes: [ ], [ 1 ]
- Utilizando un par de corchetes con objetos separados por comas: [ 1, 2, 3 ]
```

A diferencia del string la lista es una secuencia de elementos.  
Cada elemento de la lista es mutable, es decir que se pueden cambiarlo o modificarlo en cualquier momento.

Esta es una lista homogénea es decir que son del mismo tipo.

In [ ]:
Accesorios = ['Cerraduras eléctricas', 'Piloto automático', 'Asientos de cuero', 'Aire acondicionado', 'Sensor de estacionamiento', 'Sensor de lluvia']
Accesorios

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
type(Accesorios)

list

### Lista con tipos de datos variados ("Heterogéneas")

In [ ]:
Carro_1 = ['Jetta Variant', 'Motor 4.0 Turbo', 2003, 44410.0, False, ['Cerraduras eléctricas', 'Piloto automático'], 88078.64]
Carro_2 = ['Passat', 'Motor Diesel', 1991, 5712.0, False, ['Central multimedia', 'Frenos ABS'], 106161.94]

In [ ]:
Carro_1

['Jetta Variant',
 'Motor 4.0 Turbo',
 2003,
 44410.0,
 False,
 ['Cerraduras eléctricas', 'Piloto automático'],
 88078.64]

In [ ]:
Carro_2

['Passat',
 'Motor Diesel',
 1991,
 5712.0,
 False,
 ['Central multimedia', 'Frenos ABS'],
 106161.94]

***Una lista está compuesta de elementos y una lista es un elemento.***

In [ ]:
Carros = [Carro_1, Carro_2]

In [ ]:
Carros

[['Jetta Variant',
  'Motor 4.0 Turbo',
  2003,
  44410.0,
  False,
  ['Cerraduras eléctricas', 'Piloto automático'],
  88078.64],
 ['Passat',
  'Motor Diesel',
  1991,
  5712.0,
  False,
  ['Central multimedia', 'Frenos ABS'],
  106161.94]]

# 3.2 Operaciones con listas

https://docs.python.org/3.6/library/stdtypes.html#common-sequence-operations

## *x in A*

Devuelve **True** si un elemento de la lista *A es igual a x*.

*(Busca el elemento x dentro de la lista A).*

*(El elemento x pueden ser caracteríticas específicas)*

In [ ]:
Accesorios

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
'Cerraduras eléctricas' in Accesorios

True

In [ ]:
'4 x 4' in Accesorios

False

In [ ]:
'Cerraduras eléctricas' not in Accesorios

False

In [ ]:
'4 x 4' not in Accesorios

True

## *A + B*

Concatena las listas *A y B*.

In [ ]:
1 + 1

2

In [ ]:
'1' + '1'

'11'

In [ ]:
A = ['Cerraduras eléctricas', 'Piloto automático', 'Asientos de cuero']
B = ['Aire acondicionado', 'Sensor de estacionamiento', 'Sensor de lluvia']

In [ ]:
A + B

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

## *len(A)*

Tamaño de la lista A.

In [ ]:
len(Accesorios)

6

# 3.3 Selecciones en listas

## *A[ n ]*

Devuelve el n-ésimo objeto de la lista *A*.

<font color=red>**Observación:**</font> Listas tienen indexación con inicio en cero.

Esto quiere decir que el primer elemento tiene como índice 0 el segundo elemento índice 1 y así suscesivamente.

In [ ]:
Accesorios

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

Extrayendo el primer elemento.

In [ ]:
Accesorios[0]

'Cerraduras eléctricas'

In [ ]:
Accesorios[1]

'Piloto automático'

In [ ]:
Accesorios[5]

'Sensor de lluvia'

Para conocer el último elemento de nuestra lista utilizamos -1 como índice.  

Para conocer el penúltimo elemento de nuestra lista utilizamos -2 como índice.  
Para conocer el antepenúltimo elemento de nuestra lista utilizamos -3 como índice y así sucesivamente hasta llegar al primer elemento.


In [ ]:
Accesorios[-1]

'Sensor de lluvia'

In [ ]:
Carros

[['Jetta Variant',
  'Motor 4.0 Turbo',
  2003,
  44410.0,
  False,
  ['Cerraduras eléctricas', 'Piloto automático'],
  88078.64],
 ['Passat',
  'Motor Diesel',
  1991,
  5712.0,
  False,
  ['Central multimedia', 'Frenos ABS'],
  106161.94]]

In [ ]:
Carros[0]

['Jetta Variant',
 'Motor 4.0 Turbo',
 2003,
 44410.0,
 False,
 ['Cerraduras eléctricas', 'Piloto automático'],
 88078.64]

In [ ]:
Carros[1]

['Passat',
 'Motor Diesel',
 1991,
 5712.0,
 False,
 ['Central multimedia', 'Frenos ABS'],
 106161.94]

Para hallar el elemento uno dentro de la lista Carros.

In [ ]:
Carros[0][0]

'Jetta Variant'

In [ ]:
Carros[0][5]

['Cerraduras eléctricas', 'Piloto automático']

Hallando el penúltimo elemento, que es una lista.

In [ ]:
Carros[0][-2]

['Cerraduras eléctricas', 'Piloto automático']

Hallando el segundo elemento del penúltimo elemento de la lista Carros.

In [ ]:
Carros[0][-2][1]

'Piloto automático'

**Recortar Lista  (Rango de Lista)**


## *A[ i : j ]*

Recorta la lista A desde el índice i hasta j. En este recorte el elemento con índice inicial i **es incluído** y el elemento con índice final j del rango de la lista, **no es incluído** en el resultado.

In [ ]:
Accesorios

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

Entonces toma desde el segundo elemento hasta el cuarto elemento, el quinto elemento no se incluye, así es la regla.

In [ ]:
Accesorios[2:5]

['Asientos de cuero', 'Aire acondicionado', 'Sensor de estacionamiento']

Para obtener desde el segundo elemento hasta el quinto elemento.

In [ ]:
Accesorios[2:6]

['Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

O sino también de esta forma.

In [ ]:
Accesorios[2:]

['Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios[:5]

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento']

# 3.4 Métodos en listas

https://docs.python.org/3.6/library/stdtypes.html#mutable-sequence-types

## *A.sort()*

Ordena la lista *A*.  

Ordena la lista A en forma alfabética.

In [ ]:
Accesorios

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios.sort()
Accesorios

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Piloto automático',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

## *A.append(x)*

Añade el elemento *x al final de la lista A*  

Inserta un nuevo elemento al final de nuestra lista.

In [ ]:
Accesorios.append('4 x 4')
Accesorios

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Piloto automático',
 'Sensor de estacionamiento',
 'Sensor de lluvia',
 '4 x 4']

## *A.pop(i)*

Elimina y devuelve el elemento de índice i de la lista *A*.

<font color=red>**Observación:**</font> Por *defecto* el método *pop()* elimina y devuelve el último elemento de una lista.

In [ ]:
Accesorios.pop()

'4 x 4'

In [ ]:
Accesorios

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Piloto automático',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios.pop(3)

'Piloto automático'

## *A.copy()*

Crea una copia de la lista *A*.

<font color=red>**Observación:**</font> El mismo resultado puede ser obtenido con el siguiente código:
```
A[:]
```

Esta no es la forma correcta de copiar listas.

In [ ]:
Accesorios

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios_2 = Accesorios

In [ ]:
Accesorios_2

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios_2.append('4 x 4')

In [ ]:
Accesorios_2

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia',
 '4 x 4']

In [ ]:
Accesorios

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia',
 '4 x 4']

Eliminamos el elemento agregado

In [ ]:
Accesorios_2.pop()

'4 x 4'

Tanto Accesorios como Accesorios 2 tienen lo mismos elementos y si se modifica en uno se modifica en el otro. No es una variable independiente. Es la misma lista con dos nombre diferentes.

Utilizar el comando Accesorios_2 = Accesorios para crear una copia de Accesorios no es la forma correcta.

In [ ]:
Accesorios

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios_2

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

Esta es la forma correcta de copiar listas.

In [ ]:
Accesorios_2 = Accesorios.copy()

In [ ]:
Accesorios_2

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios_2.append('4 x 4')
Accesorios_2

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia',
 '4 x 4']

In [ ]:
Accesorios_2.pop()
Accesorios_2

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
Accesorios

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

Esto es lo mismo **Accesorios_2 = Accesorios[:]** que efectuar el método copy **Accesorios_2 = Accesorios.copy()**

In [ ]:
# Accesorios_2 = Accesorios[:]
# Accesorios_2 = Accesorios.copy()

In [ ]:
Accesorios_2 = Accesorios[:]
Accesorios_2

['Aire acondicionado',
 'Asientos de cuero',
 'Cerraduras eléctricas',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

# <font color=green> 4. ESTRUCTURAS DE REPETICIÓN Y CONDICIONALES
---

# 4.1 Instrucción *for*

#### Formato estándar  

Iterar o recorrer  

coleccion = conjunto de elementos que puede ser una lista

```
for <variable> in <coleccion>:
    <instrucciones>
```

### Loops con listas

In [ ]:
Accesorios = ['Cerraduras eléctricas', 'Piloto automático', 'Asientos de cuero', 'Aire acondicionado', 'Sensor de estacionamiento', 'Sensor de lluvia']
Accesorios

['Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de lluvia']

In [ ]:
for item in Accesorios:
  print(item)

Cerraduras eléctricas
Piloto automático
Asientos de cuero
Aire acondicionado
Sensor de estacionamiento
Sensor de lluvia


Se observa que el print obtenido no es una lista sino un print de los elementos de la lista accesorios. Es decir no devuelve una lista.

###  List comprehensions

https://docs.python.org/3.6/tutorial/datastructures.html#list-comprehensions

*range()* -> https://docs.python.org/3.6/library/functions.html#func-range

List comprehensions es una forma de crear listas pero más compacta en lugar de utilizar 2 o más líneas para crear una lista.

el método range() me devuelve un rango de elementos determinados.

In [ ]:
range(10)

range(0, 10)

Creo una lista utilizando range y en range le asigno la cantidad de elementos de la lista.

In [ ]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [ ]:
for i in range(10):
  print(i ** 2)

0
1
4
9
16
25
36
49
64
81


In [ ]:
for i in range(10):
  print(list)

<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>


Utilizando una lista vacía [ ] y le creamos elementos para guardalos dentro.

In [ ]:
cuadrado = []
for i in range(10):
  cuadrado.append(i**2)
cuadrado

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

List comprehensions

In [ ]:
cuadrado=[i**2 for i in range(10)]

In [ ]:
cuadrado

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

# 4.2 Loops anidados

Utilizar un for dentro de otro for

La lista datos tiene como elementos 3 listas cada una con 8 elementos.

In [ ]:
datos = [
    ['Llantas de aleación', 'Cerraduras eléctricas', 'Piloto automático', 'Asientos de cuero', 'Aire acondicionado', 'Sensor de estacionamiento', 'Sensor de crepúsculo', 'Sensor de lluvia'],
    ['Central multimedia', 'Techo panorámico', 'Frenos ABS', '4 X 4', 'Panel digital', 'Piloto automático', 'Asientos de cuero', 'Cámara de estacionamiento'],
    ['Piloto automático', 'Control de estabilidad', 'Sensor crepuscular', 'Frenos ABS', 'Transmisión automática', 'Asientos de cuero', 'Central multimedia', 'Ventanas eléctricas']
]
datos

[['Llantas de aleación',
  'Cerraduras eléctricas',
  'Piloto automático',
  'Asientos de cuero',
  'Aire acondicionado',
  'Sensor de estacionamiento',
  'Sensor de crepúsculo',
  'Sensor de lluvia'],
 ['Central multimedia',
  'Techo panorámico',
  'Frenos ABS',
  '4 X 4',
  'Panel digital',
  'Piloto automático',
  'Asientos de cuero',
  'Cámara de estacionamiento'],
 ['Piloto automático',
  'Control de estabilidad',
  'Sensor crepuscular',
  'Frenos ABS',
  'Transmisión automática',
  'Asientos de cuero',
  'Central multimedia',
  'Ventanas eléctricas']]

Con este primer for llego a las listas. **¿Cómo llego a los elementos de cada sub-lista?**

***Observando vemos que cada elemento de mi lista es una lista y cada lista es una colección de datos***

In [ ]:
for lista in datos:
  print(lista)

['Llantas de aleación', 'Cerraduras eléctricas', 'Piloto automático', 'Asientos de cuero', 'Aire acondicionado', 'Sensor de estacionamiento', 'Sensor de crepúsculo', 'Sensor de lluvia']
['Central multimedia', 'Techo panorámico', 'Frenos ABS', '4 X 4', 'Panel digital', 'Piloto automático', 'Asientos de cuero', 'Cámara de estacionamiento']
['Piloto automático', 'Control de estabilidad', 'Sensor crepuscular', 'Frenos ABS', 'Transmisión automática', 'Asientos de cuero', 'Central multimedia', 'Ventanas eléctricas']


Colocamos otro for y a cada elemento lo llamamos item y ahora que vamos recorrer no es la lista Datos sino cada una de las sub-listas y como hemos llamado a estas sub-listas, la hemos llamado lista (Que es una colección).

In [ ]:
for lista in datos:
  for item in lista:
    print(item)

Llantas de aleación
Cerraduras eléctricas
Piloto automático
Asientos de cuero
Aire acondicionado
Sensor de estacionamiento
Sensor de crepúsculo
Sensor de lluvia
Central multimedia
Techo panorámico
Frenos ABS
4 X 4
Panel digital
Piloto automático
Asientos de cuero
Cámara de estacionamiento
Piloto automático
Control de estabilidad
Sensor crepuscular
Frenos ABS
Transmisión automática
Asientos de cuero
Central multimedia
Ventanas eléctricas


Ahora sí, hemos llegado a los elementos de las sub-listas.

**Ahora vamos a guardar estos elementos en una variable que va a ser una lista.**

In [ ]:
Accesorios = []

for lista in datos:
  for item in lista:
    Accesorios.append(item)
Accesorios

['Llantas de aleación',
 'Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de crepúsculo',
 'Sensor de lluvia',
 'Central multimedia',
 'Techo panorámico',
 'Frenos ABS',
 '4 X 4',
 'Panel digital',
 'Piloto automático',
 'Asientos de cuero',
 'Cámara de estacionamiento',
 'Piloto automático',
 'Control de estabilidad',
 'Sensor crepuscular',
 'Frenos ABS',
 'Transmisión automática',
 'Asientos de cuero',
 'Central multimedia',
 'Ventanas eléctricas']

In [ ]:
len(Accesorios)

24

Y ya, tenemos creada la lista Accesorios con elementos independientes, individuales ya no hay más sub-listas, pero hay elementos que se repiten.

Ahora es más fácil de usar en un modelo de machine learning.

## *set()*

https://docs.python.org/3.6/library/stdtypes.html#types-set

https://docs.python.org/3.6/library/functions.html#func-set

Método set() para remover elementos duplicados de una colección o lista, pero no devuelve una lista

Y ordena alfabéticamente los elementos.

In [ ]:
set(Accesorios)

{'4 X 4',
 'Aire acondicionado',
 'Asientos de cuero',
 'Central multimedia',
 'Cerraduras eléctricas',
 'Control de estabilidad',
 'Cámara de estacionamiento',
 'Frenos ABS',
 'Llantas de aleación',
 'Panel digital',
 'Piloto automático',
 'Sensor crepuscular',
 'Sensor de crepúsculo',
 'Sensor de estacionamiento',
 'Sensor de lluvia',
 'Techo panorámico',
 'Transmisión automática',
 'Ventanas eléctricas'}

In [ ]:
len(set(Accesorios))

18

Entonces para convertirlo en una lista:

In [ ]:
list(set(Accesorios))

['Aire acondicionado',
 'Central multimedia',
 'Asientos de cuero',
 'Ventanas eléctricas',
 'Panel digital',
 'Frenos ABS',
 'Sensor crepuscular',
 'Transmisión automática',
 'Cerraduras eléctricas',
 '4 X 4',
 'Control de estabilidad',
 'Cámara de estacionamiento',
 'Piloto automático',
 'Sensor de crepúsculo',
 'Sensor de lluvia',
 'Techo panorámico',
 'Llantas de aleación',
 'Sensor de estacionamiento']

### List comprehensions

**Lo utilizamos cada vez que queremos crear listas**

In [ ]:
Accesorios = [item for lista in datos for item in lista]

In [ ]:
Accesorios

['Llantas de aleación',
 'Cerraduras eléctricas',
 'Piloto automático',
 'Asientos de cuero',
 'Aire acondicionado',
 'Sensor de estacionamiento',
 'Sensor de crepúsculo',
 'Sensor de lluvia',
 'Central multimedia',
 'Techo panorámico',
 'Frenos ABS',
 '4 X 4',
 'Panel digital',
 'Piloto automático',
 'Asientos de cuero',
 'Cámara de estacionamiento',
 'Piloto automático',
 'Control de estabilidad',
 'Sensor crepuscular',
 'Frenos ABS',
 'Transmisión automática',
 'Asientos de cuero',
 'Central multimedia',
 'Ventanas eléctricas']

In [ ]:
len(Accesorios)

24

In [ ]:
Accesorios = list(set([item for lista in datos for item in lista]))

In [ ]:
Accesorios

['Aire acondicionado',
 'Central multimedia',
 'Asientos de cuero',
 'Ventanas eléctricas',
 'Panel digital',
 'Frenos ABS',
 'Sensor crepuscular',
 'Transmisión automática',
 'Cerraduras eléctricas',
 '4 X 4',
 'Control de estabilidad',
 'Cámara de estacionamiento',
 'Piloto automático',
 'Sensor de crepúsculo',
 'Sensor de lluvia',
 'Techo panorámico',
 'Llantas de aleación',
 'Sensor de estacionamiento']

In [ ]:
len(Accesorios)

18

# 4.3 Instrucción *if*

o Bucle if: se utiliza para comparaciones, diferencias, mayor que, menor que, igual mayor que, igual menor que, además de los operadores lógicos.

#### Formato estándar

```
if <condicion>:
     <instrucciones en caso la condición sea verdadera>
```

### Operadores de comparación: $==$, $!=$, $>$, $<$, $>=$, $<=$
### y
### Operadores lógicos: $and$, $or$, $not$

In [ ]:
# 1º objeto o elemento de la lista - Nombre del vehículo  (Este es un string)
# 2º objeto o elemento de la lista - Año de fabricación   (Este es un número entero)
# 3º objeto o elemento de la lista - Vehículo es cero km? (Este es un booleano)

datos = [
    ['Jetta Variant', 2003, False],
    ['Passat', 1991, False],
    ['Crossfox', 1990, False],
    ['DS5', 2019, True],
    ['Aston Martin DB4', 2006, False],
    ['Palio Weekend', 2012, False],
    ['A5', 2019, True],
    ['Série 3 Cabrio', 2009, False],
    ['Dodge Jorney', 2019, False],
    ['Carens', 2011, False]
]
datos

[['Jetta Variant', 2003, False],
 ['Passat', 1991, False],
 ['Crossfox', 1990, False],
 ['DS5', 2019, True],
 ['Aston Martin DB4', 2006, False],
 ['Palio Weekend', 2012, False],
 ['A5', 2019, True],
 ['Série 3 Cabrio', 2009, False],
 ['Dodge Jorney', 2019, False],
 ['Carens', 2011, False]]

Extrayendo los vehículos de la lista datos los autos que tienen cero kilometros.
Si en la sub-lista es True quiere decir que el vehículo es cero kilometros y es el elemento que quiero extraer.

Para lo cual recorremos la lista datos para posicionarmen en cada sub-lista hacemos esto con un for denominando la variable lista con el índice del elemento que queremos extraer y la lista a recorrer es datos, luego para comparar utilizamos la sentencia if y la condición es la comparación de una cosa con otra cosa.

In [ ]:
for lista in datos:
  if (lista[2] == True):
    print(lista)


['DS5', 2019, True]
['A5', 2019, True]


Ahora para que no se quede en pantalla guardamos la función en la variable cero_kilometros. Y para agregar los elementos resultantes a la lista vacía cero_km_t utilizamos el método append(). Y dentro de append colocamos la variable lista. La cual nos devuelve una lista con dos sub-listas como elementos.

In [ ]:
cero_km_T = []

for lista in datos:
  if (lista[2] == True):
    cero_km_T.append(lista)
cero_km_T

[['DS5', 2019, True], ['A5', 2019, True]]

In [ ]:
len(cero_km_T)

2

Ahora queremos ver los autos que no son cero kilometros y lo guardamos en la lista cero_km_False

In [ ]:
cero_km_F = []

for lista in datos:
  if (lista[2] == False):
    cero_km_F.append(lista)
cero_km_F

[['Jetta Variant', 2003, False],
 ['Passat', 1991, False],
 ['Crossfox', 1990, False],
 ['Aston Martin DB4', 2006, False],
 ['Palio Weekend', 2012, False],
 ['Série 3 Cabrio', 2009, False],
 ['Dodge Jorney', 2019, False],
 ['Carens', 2011, False]]

In [ ]:
len(cero_km_F)

8

### List comprehensions

In [ ]:
[lista for lista in datos if (lista[2] == True) ]

[['DS5', 2019, True], ['A5', 2019, True]]

In [ ]:
cero_km = [lista for lista in datos if (lista[2] == True) ]
cero_km

[['DS5', 2019, True], ['A5', 2019, True]]

# 4.4 Instrucciones *if-else* y *if-elif-else*

#### Formato estándar

```
if <condicion>:
    <instrucciones en caso la condición sea verdadera>
else:
    <instrucciones en caso la condición no sea verdadera>
```

In [ ]:
cero_km_T, cero_km_F  = [] , []

for lista in datos:
  if (lista[2] == True):
    cero_km_T.append(lista)
  else:
    cero_km_F.append(lista)


In [ ]:
cero_km_T

[['DS5', 2019, True], ['A5', 2019, True]]

In [ ]:
type(cero_km_T)

list

In [ ]:
len(cero_km_T)

2

In [ ]:
cero_km_F

[['Jetta Variant', 2003, False],
 ['Passat', 1991, False],
 ['Crossfox', 1990, False],
 ['Aston Martin DB4', 2006, False],
 ['Palio Weekend', 2012, False],
 ['Série 3 Cabrio', 2009, False],
 ['Dodge Jorney', 2019, False],
 ['Carens', 2011, False]]

In [ ]:
type(cero_km_F)

list

In [ ]:
len(cero_km_F)

8

### Formato estándar

```
if <condicion 1>:
    <instrucciones en caso la condición 1 sea verdadera>
elif <condicion 2>:
    <instrucciones en caso la condición 2 sea verdadera>
elif <condicion 3>:
    <instrucciones en caso la condición 3 sea verdadera>
                        .
                        .
                        .
else:
    <instrucciones en caso las condiciones anteriores no sean verdaderas>
```

else va a realizar todas las condiciones que no se cumplen en las condiciones anteriores y seguir las instrucciones que agrupen a esos elementos.

In [ ]:
datos

[['Jetta Variant', 2003, False],
 ['Passat', 1991, False],
 ['Crossfox', 1990, False],
 ['DS5', 2019, True],
 ['Aston Martin DB4', 2006, False],
 ['Palio Weekend', 2012, False],
 ['A5', 2019, True],
 ['Série 3 Cabrio', 2009, False],
 ['Dodge Jorney', 2019, False],
 ['Carens', 2011, False]]

In [ ]:
len(datos)

10

In [ ]:
print('AND')
print(f'(True and True) el resultado es: {True and True}')
print(f'(True and False) el resultado es: {True and False}')
print(f'(False and True) el resultado es: {False and True}')
print(f'(False and False) el resultado es: {False and False}')

AND
(True and True) el resultado es: True
(True and False) el resultado es: False
(False and True) el resultado es: False
(False and False) el resultado es: False


In [ ]:
print('OR')
print(f'(True or True) el resultado es: {True or True}')
print(f'(True or False) el resultado es: {True or False}')
print(f'(False or True) el resultado es: {False or True}')
print(f'(False or False) el resultado es: {False or False}')

OR
(True or True) el resultado es: True
(True or False) el resultado es: True
(False or True) el resultado es: True
(False or False) el resultado es: False


In [ ]:
A, B, C = [] , [] , []

for lista in datos:
  if(lista[1] <= 2000):
    A.append(lista)
  elif(lista[1] > 2000 and lista[1] <= 2010):
    B.append(lista)
  else:
    C.append(lista)

In [ ]:
A

[['Passat', 1991, False], ['Crossfox', 1990, False]]

In [ ]:
B

[['Jetta Variant', 2003, False],
 ['Aston Martin DB4', 2006, False],
 ['Série 3 Cabrio', 2009, False]]

In [ ]:
C

[['DS5', 2019, True],
 ['Palio Weekend', 2012, False],
 ['A5', 2019, True],
 ['Dodge Jorney', 2019, False],
 ['Carens', 2011, False]]

Otra forma de hacer la función

In [ ]:
A, B, C = [] , [] , []

for lista in datos:
  if(lista[1] <= 2000):
    A.append(lista)
  elif(2000 < lista[1] <= 2010):
    B.append(lista)
  else:
    C.append(lista)

In [ ]:
A

[['Passat', 1991, False], ['Crossfox', 1990, False]]

In [ ]:
B

[['Jetta Variant', 2003, False],
 ['Aston Martin DB4', 2006, False],
 ['Série 3 Cabrio', 2009, False]]

In [ ]:
C

[['DS5', 2019, True],
 ['Palio Weekend', 2012, False],
 ['A5', 2019, True],
 ['Dodge Jorney', 2019, False],
 ['Carens', 2011, False]]

# <font color=green> 5. NUMPY BÁSICO
---

Numpy es la abreviación de **Numerical Python** y es uno de los paquetes más importantes para procesamiento numérico en Python. **Numpy ofrece la base para la mayoría de los paquetes de aplicaciones científicas que usen datos numéricos en Python (estructuras de datos y algoritmos).** Pueden destacarse los siguientes recursos que el paquete Numpy contiene:

- Un poderoso objeto array multidimensional;
- Funciones matemáticas sofisticadas para operaciones con arrays sin la necesidad de utilización de bucles *for*;
- Recursos de algebra linear y generación de números aleatórios

Además de sus obvios usos científicos, el paquete NumPy también es muy utilizado en análisis de datos como un eficiente contenedor multidimensional de datos genéricos para transporte entre diversos algoritmos y bibliotecas en Python.

**Versión:** 1.18.4

**Instalación:** https://scipy.org/install.html

**Documentación:** https://numpy.org/doc/1.18/

### Paquetes

Existen diversos paquetes Python disponibles para download en internet. Cada paquete tiene como objetivo la solución de determinado tipo de problema y para esto son desarrollados nuevos tipos, funciones y métodos.

Algunos paquetes son bastante utilizados en un contexto de ciencia de datos, como por ejemplo:

- Numpy
- Pandas
- Scikit-learn
- Matplotlib

Algunos paquetes no son distribuidos en la instalación estándar de Python. En este caso debemos instalar los paquetes que necesitamos en nuestro sistema para poder utilizar sus funcionalidades.

### Importando todo el paquete con su nombre completo

In [ ]:
import numpy


In [ ]:
numpy.__version__

'1.23.5'

https://numpy.org/doc/1.16/reference/generated/numpy.arange.html

In [ ]:
numpy.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

Array en programación: qué es
El array, dentro de la programación, se refiere a un tipo de dato estructurado que almacena datos, del mismo tipo y relacionados. Los elementos que conforman pueden ser de tipo simple o compuesto, como en el caso de vectores o listas.

Los datos que almacena el array son conocidos como elementos y, al número de estos elementos que hay dentro del array, se les denomina rango del vector o tamaño. Asimismo, para acceder a ellos, se utiliza un índice. Es decir, un número que indica la posición del elemento concreto ubicado en el array.

### Importando todo el paquete y asignándole un nuevo nombre

In [ ]:
import numpy as np


In [ ]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

### Importando parte del paquete

In [ ]:
from numpy import arange

In [ ]:
arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

# 5.1 Creando arrays Numpy

Utilizando el método array()

In [ ]:
import numpy as np

### A partir de listas

https://numpy.org/doc/1.16/user/basics.creation.html

Convirtiendo la variable km en un array de numpy

In [ ]:
km = np.array([1000, 2300, 4987, 1500])

In [ ]:
km

array([1000, 2300, 4987, 1500])

In [ ]:
type(km)

numpy.ndarray

https://numpy.org/doc/1.16/user/basics.types.html

Empleando el método dtype() = datos tipo; tipo de datos (elementos) que contiene el array

En este caso números enteros.

In [ ]:
km.dtype

dtype('int64')

### A partir de datos externos

https://numpy.org/doc/1.16/reference/generated/numpy.loadtxt.html

In [ ]:
km = np.loadtxt(fname='/content/drive/MyDrive/data/carros-km.txt', dtype=int)

In [ ]:
km

array([ 44410,   5712,  37123,      0,  25757,  10728,      0,  77599,
        99197,  37978,  12859,   8052,  89773,      0,  41457, 115607,
        46449,      0,  37086,  15173, 101193,      0,  98079, 102959,
            0,      0,   5795,      0,  58848,  94381,  30163,  53332,
        17720,  33808,  90684,  43975,      0,      0,   5526,      0,
        93415,  40762,      0,  86302,      0,   9755,  69945,   2395,
            0,  80349,  85554,  50496,  67716,  93947,  35345,  81007,
       119513,      0,      0,      0,      0,      0, 118895,  48509,
       100912,  95649,      0,  90495,      0,  29132,  23802,  84992,
        54395,  26731,  44329, 118236, 113808,    610,      0,      0,
        12887,  79607,  90924,  42733,      0,      0, 117714, 113885,
            0,  30511,  74867, 119760,   8356,  64247,  88661,   4539,
       110116,  33215,  92001,      0,  81708,  70641,      0,  91277,
        26544,  52596,  47503,  89056,  28834, 110564,  56638,  17357,
      

Como vemos nos origino un array unidimensional.

In [ ]:
km.dtype

dtype('int64')

### Arrays con dos dimensiones

In [ ]:
datos = [
    ['Llantas de aleación', 'Cerraduras eléctricas', 'Piloto automático', 'Asientos de cuero', 'Aire acondicionado', 'Sensor de estacionamiento', 'Sensor de crepúsculo', 'Sensor de lluvia'],
    ['Central multimedia', 'Techo panorámico', 'Frenos ABS', '4 X 4', 'Panel digital', 'Piloto automático', 'Asientos de cuero', 'Cámara de estacionamiento'],
    ['Piloto automático', 'Control de estabilidad', 'Sensor crepuscular', 'Frenos ABS', 'Transmisión automática', 'Asientos de cuero', 'Central multimedia', 'Ventanas eléctricas']
]
datos

[['Llantas de aleación',
  'Cerraduras eléctricas',
  'Piloto automático',
  'Asientos de cuero',
  'Aire acondicionado',
  'Sensor de estacionamiento',
  'Sensor de crepúsculo',
  'Sensor de lluvia'],
 ['Central multimedia',
  'Techo panorámico',
  'Frenos ABS',
  '4 X 4',
  'Panel digital',
  'Piloto automático',
  'Asientos de cuero',
  'Cámara de estacionamiento'],
 ['Piloto automático',
  'Control de estabilidad',
  'Sensor crepuscular',
  'Frenos ABS',
  'Transmisión automática',
  'Asientos de cuero',
  'Central multimedia',
  'Ventanas eléctricas']]

In [ ]:
len(datos)

3

In [ ]:
Accesorios = np.array(datos)

In [ ]:
Accesorios

array([['Llantas de aleación', 'Cerraduras eléctricas',
        'Piloto automático', 'Asientos de cuero', 'Aire acondicionado',
        'Sensor de estacionamiento', 'Sensor de crepúsculo',
        'Sensor de lluvia'],
       ['Central multimedia', 'Techo panorámico', 'Frenos ABS', '4 X 4',
        'Panel digital', 'Piloto automático', 'Asientos de cuero',
        'Cámara de estacionamiento'],
       ['Piloto automático', 'Control de estabilidad',
        'Sensor crepuscular', 'Frenos ABS', 'Transmisión automática',
        'Asientos de cuero', 'Central multimedia', 'Ventanas eléctricas']],
      dtype='<U25')

In [ ]:
km.shape

(258,)

In [ ]:
Accesorios.shape

(3, 8)

### Comparando desempeño de arrays con listas

In [ ]:
np_array = np.arange(1000000)

In [ ]:
py_list = list(range(1000000))

In [ ]:
%time for _ in range(100): np_array *= 2

CPU times: user 115 ms, sys: 3.95 ms, total: 119 ms
Wall time: 166 ms


In [ ]:
%time for _ in range(100): py_list = [x*2 for x in py_list]

CPU times: user 9.93 s, sys: 2.13 s, total: 12.1 s
Wall time: 12.4 s


# 5.2 Operaciones aritméticas con arrays Numpy

### Operaciones entre arrays y constantes

**Quiero obtener el número de años(edad) de los vehículos.

In [ ]:
km = [44410., 5712., 37123., 0., 25757.]
años = [2003, 1991, 1990, 2019, 2006]

No funciona porque no se puede mezclar enteros con listas, es decir el entero 2023 con la lista años.

In [ ]:
# edad = 2023- años

Para esto es mejor utilizar arrays numpy

In [ ]:
km = np.array([44410., 5712., 37123., 0., 25757.])
años = np.array([2003, 1991, 1990, 2019, 2006])

In [ ]:
edad = 2023- años

In [ ]:
edad

array([20, 32, 33,  4, 17])

In [ ]:
km_media = km / edad

In [ ]:
km_media

array([2220.5       ,  178.5       , 1124.93939394,    0.        ,
       1515.11764706])

Comprobando

In [ ]:
44410/20

2220.5

In [ ]:
5712/32

178.5

In [ ]:
37123/33

1124.939393939394

In [ ]:
0/4

0.0

In [ ]:
25757/17

1515.1176470588234

### Operaciones entre arrays

Creando un array de dos dimensiones llamado datos.

In [ ]:
datos = np.array([km, años])

Como vemos datos tiene dos dimensiones

In [ ]:
datos

array([[44410.,  5712., 37123.,     0., 25757.],
       [ 2003.,  1991.,  1990.,  2019.,  2006.]])

In [ ]:
type(datos)

numpy.ndarray

In [ ]:
datos.dtype

dtype('float64')

In [ ]:
datos.shape

(2, 5)

Entonces tenemos dos filas y 5 columnas en el array datos.

### Operaciones con arrays de dos dimensiones

**Nota.-** del array datos las 2 filas(horizontal) la representamos con índice 0 y 1, las columnas (vertical) las representamos con los índices 0,1,2,3,4.***(Como se ve en la imagen).***

Primera dimensión

In [ ]:
datos[0]

array([44410.,  5712., 37123.,     0., 25757.])

Segunda dimensión

In [ ]:
datos[1]

array([2003., 1991., 1990., 2019., 2006.])

Entonces ahora tenemos separados nuestras dos dimensiones

In [ ]:
km_media = datos[0] / (2023 - datos[1])
km_media

array([2220.5       ,  178.5       , 1124.93939394,    0.        ,
       1515.11764706])

Hemos generado la misma ecuación utilizando un solo array numpy, accesando separadamente a sus dos dimensiones.

Obteniendo los mismo resultados.

![1410-img01.png](https://caelum-online-public.s3.amazonaws.com/1410-pythondatascience/01/1410-img01.png)

# 5.3 Selecciones con arrays Numpy

![1410-img01.png](https://caelum-online-public.s3.amazonaws.com/1410-pythondatascience/01/1410-img01.png)

In [ ]:
datos

array([[44410.,  5712., 37123.,     0., 25757.],
       [ 2003.,  1991.,  1990.,  2019.,  2006.]])

Cada columna representa un vehícula y cada fila un tipo de información


![1410-img02.png](https://caelum-online-public.s3.amazonaws.com/1410-pythondatascience/01/1410-img02.png)

### Indexación

<font color=red>**Observación:**</font> La indexación tiene inicio en cero.

In [ ]:
contador = np.arange(10)
contador

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
contador[0]

0

In [ ]:
contador[9]

9

In [ ]:
contador[-1]

9

In [ ]:
datos[0]

array([44410.,  5712., 37123.,     0., 25757.])

In [ ]:
datos[1]

array([2003., 1991., 1990., 2019., 2006.])

Quiero extraer el valor 1990 del array datos.  
Para realizar esto, primero debemos determinar su ubicación en las filas y luego su ubicación en las columnas.

## <font color=green>**Consejo:**</font>
### *ndarray[ linea ][ columna ]* - o sino -  *ndarray[ linea, columna ]*

In [ ]:
datos[1] [2]

1990.0

In [ ]:
datos[1,2]

1990.0

**Ahora vamos a aprender a seleccionar un subconjunto de elementos que puede ser un subconjunto de líneas o un subconjunto de columnas o la intersección entre ellas.**

 ### Particiones

La sintaxis para realizar particiones en un array Numpy es $i : j : k$ donde $i$ es el índice inicial, $j$ es el índice final, y $k$ es el indicador de intervalo ($k\neq0$)

<font color=red>**Observación:**</font> En las particiones (*slices*) el objeto con índice i es **incluído** y el objeto con índice j **no es incluído** en el resultado.

![1410-img01.png](https://caelum-online-public.s3.amazonaws.com/1410-pythondatascience/01/1410-img01.png)

Para un array unidimensional utilizamos la fórmula i:j:k
Para arrays bidemesionales usamos la fórmula [línea, columna]

In [ ]:
contador = np.arange(10)
contador

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

Array unidimensional: formato i : j : k

i para seleccionar el primer elemento

j para seleccionar el último elemento

k indicador de intervalo

In [ ]:
contador[1:4:]

array([1, 2, 3])

In [ ]:
contador[1:8:2]

array([1, 3, 5, 7])

Selecciono el elemento uno y el elemento final 7, y el intervalo entre estos saltando 2 posiciones.

Ahora quiero los números pares.

Que significa que deje en blanco la primer y segunda condición, que le estamos indicando al interpretador de python que tome el primer y el último elemento del array. Y para hallar los números pares que salte 2 posiciones.

In [ ]:
contador[ : : 2]

array([0, 2, 4, 6, 8])

Ahora quiero los números impares

In [ ]:
contador[1 : : 2]

array([1, 3, 5, 7, 9])

In [ ]:
datos

array([[44410.,  5712., 37123.,     0., 25757.],
       [ 2003.,  1991.,  1990.,  2019.,  2006.]])

**Particiones bidimensionales**

array[linea, columna]

In [ ]:
datos[ : : , 1 : 3 : ]

array([[ 5712., 37123.],
       [ 1991.,  1990.]])

Calcular el promedio de kilometraje de estos dos vehículos.
km_promedio = kilometraje/(2023 - año de fabricación)

In [ ]:
datos[ : : , 1 : 3 : ][0]

array([ 5712., 37123.])

In [ ]:
datos[ : : , 1 : 3 : ][1]

array([1991., 1990.])

Ahora con las filas del array separadadas procedemos a calcular la fórmula.

In [ ]:
datos[ : : , 1 : 3 : ][0] / (2023 - datos[ : : , 1 : 3 : ][1])

array([ 178.5       , 1124.93939394])

Realizando el corte para todos los vehículos.

In [ ]:
datos[0] / (2023 - datos[1])

array([2220.5       ,  178.5       , 1124.93939394,    0.        ,
       1515.11764706])

### Indexación con array booleano

<font color=red>**Observación:**</font> Selecciona un grupo de líneas y columnas siguiendo las etiquetas o un array booleano.

In [ ]:
contador = np.arange(10)
contador

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
contador > 5

array([False, False, False, False, False, False,  True,  True,  True,
        True])

In [ ]:
contador[contador > 5]

array([6, 7, 8, 9])

In [ ]:
datos

array([[44410.,  5712., 37123.,     0., 25757.],
       [ 2003.,  1991.,  1990.,  2019.,  2006.]])

In [ ]:
datos[1] > 2000

array([ True, False, False,  True,  True])

Quiero seleccionar los vehículos año de fabricación mayor a 2,000.00

In [ ]:
datos[ : : , datos[1] > 2000]

array([[44410.,     0., 25757.],
       [ 2003.,  2019.,  2006.]])

# 5.4 Atributos y métodos de arrays Numpy

### Atributos:

**Son características o propiedades de un array numpy. Estos atributos no necesitan los paréntesis al final de su nombre.**

https://numpy.org/doc/1.16/reference/arrays.ndarray.html#array-attributes

In [ ]:
datos

array([[44410.,  5712., 37123.,     0., 25757.],
       [ 2003.,  1991.,  1990.,  2019.,  2006.]])

## *ndarray.shape*

Devuelve una tupla con las dimensiones del array.

*Es decir, filas a la izquierda y a la derecha columnas*

In [ ]:
datos.shape

(2, 5)

2 filas y 5 columnas.

## *ndarray.ndim*

Devuelve el número de dimensiones del array.

In [ ]:
datos.ndim

2

## *ndarray.size*

Devuelve el número de elementos del array.

In [ ]:
datos.size

10

## *ndarray.dtype*

Devuelve el tipo de datos de los elementos del array.

In [ ]:
datos.dtype

dtype('float64')

## *ndarray.T*

Devuelve el array transpuesto, es decir, convierte líneas en columnas y viceversa.

In [ ]:
datos.T

array([[44410.,  2003.],
       [ 5712.,  1991.],
       [37123.,  1990.],
       [    0.,  2019.],
       [25757.,  2006.]])

In [ ]:
datos.transpose()

array([[44410.,  2003.],
       [ 5712.,  1991.],
       [37123.,  1990.],
       [    0.,  2019.],
       [25757.,  2006.]])

### Métodos

https://numpy.org/doc/1.16/reference/arrays.ndarray.html#array-methods

**Los métodos son como funciones entonces toda función usa los paréntesis al momento de invocarlas porque dentro de estos paréntesis se pueden colocar alguna condición a usar por el método.**

## *ndarray.tolist()*

Devuelve el array como una lista Python.

In [524]:
datos.tolist()

[[44410.0, 5712.0, 37123.0, 0.0, 25757.0],
 [2003.0, 1991.0, 1990.0, 2019.0, 2006.0]]

## *ndarray.reshape(shape[, order])*

Devuelve un array que contiene los mismos datos con una nueva forma.

Devuelve los mismos valores organizados de una forma diferente.

In [525]:
contador = np.arange(10)
contador

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

.reshape((linea, columnas))

In [527]:
contador.reshape((5,2))

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7],
       [8, 9]])

Tipo de orden: es decir le puedo decir al método shape que lo ordene de una manera u otra.

Quiero decirle que este contador se ordene siguiento el lenguaje C.

In [528]:
contador.reshape((5,2), order = 'C')

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7],
       [8, 9]])

Quiero decirle que este contador se ordene siguiento el lenguaje F  (Fortran).

In [529]:
contador.reshape((5,2), order = 'F')

array([[0, 5],
       [1, 6],
       [2, 7],
       [3, 8],
       [4, 9]])

La ordenación obtenida es diferente.

Ahora con dos lista vamos a concatenarlas

In [530]:
km = [44410, 5712, 37123, 0, 25757]
años = [2003, 1991, 1990, 2019, 2006]

In [531]:
info_carros = km + años
info_carros

[44410, 5712, 37123, 0, 25757, 2003, 1991, 1990, 2019, 2006]

Modificando las listas aun array y dimensionándolas.  

np.array(info_carros).reshape(linea, columnas)

In [532]:
np.array(info_carros).reshape(2, 5)

array([[44410,  5712, 37123,     0, 25757],
       [ 2003,  1991,  1990,  2019,  2006]])

Quiero que cada información del vehículo este en una columna.

In [533]:
np.array(info_carros).reshape((5, 2) , order= 'F')

array([[44410,  2003],
       [ 5712,  1991],
       [37123,  1990],
       [    0,  2019],
       [25757,  2006]])

## *ndarray.resize(new_shape[, refcheck])*

Cambia la forma y el tamaño del array.

In [534]:
datos_new = datos.copy()

In [535]:
datos_new

array([[44410.,  5712., 37123.,     0., 25757.],
       [ 2003.,  1991.,  1990.,  2019.,  2006.]])

datos_new.resize((lineas, columnas))

ValueError: cannot resize an array that references or is referenced
by another array in this way.
Use the np.resize function or refcheck=False

In [537]:
datos_new.resize((3, 5), refcheck=False)

Nos devuelve una línea adicional con valores cero.

In [538]:
datos_new

array([[44410.,  5712., 37123.,     0., 25757.],
       [ 2003.,  1991.,  1990.,  2019.,  2006.],
       [    0.,     0.,     0.,     0.,     0.]])

Ahora para rellenar la 3ra fila con los datos promedio del kilometraje

In [539]:
datos_new[2] = datos_new[0] / (2023 - datos_new[1])

In [540]:
datos_new

array([[44410.        ,  5712.        , 37123.        ,     0.        ,
        25757.        ],
       [ 2003.        ,  1991.        ,  1990.        ,  2019.        ,
         2006.        ],
       [ 2220.5       ,   178.5       ,  1124.93939394,     0.        ,
         1515.11764706]])

# 5.5 Estadísticas con arrays Numpy

https://numpy.org/doc/1.16/reference/arrays.ndarray.html#calculation

y

https://numpy.org/doc/1.16/reference/routines.statistics.html

y

https://numpy.org/doc/1.16/reference/routines.math.html

**Cargando los archivos desde mi drive en el folder data.**

In [541]:
años = np.loadtxt(fname = "/content/drive/MyDrive/data/carros-años.txt", dtype = int)
km = np.loadtxt(fname = "/content/drive/MyDrive/data/carros-km.txt")
valor = np.loadtxt(fname = "/content/drive/MyDrive/data/carros-valor.txt")

array unidimensional o independiente.

In [542]:
años.shape

(258,)

**Para realizar cálculos estadísticos necesitamos un array bidimensional**

Para esto vamos a combinar cada uno de los 3 arrays independientes para convertirlo en una única matriz de filas y columnas que contenga estos valores.

https://numpy.org/doc/1.16/reference/generated/numpy.column_stack.html

Utilizando los archivos cargados definimos la matriz que debe tener en la primera columna los años, en la segunda columna el kilometraje y en la tercera columna el valor de los vehículos.  
Por lo que nos va a crear una matriz con 258 filas y 3 columnas.


In [543]:
dataset = np.column_stack((años, km, valor))
dataset

array([[2.0030000e+03, 4.4410000e+04, 8.8078640e+04],
       [1.9910000e+03, 5.7120000e+03, 1.0616194e+05],
       [1.9900000e+03, 3.7123000e+04, 7.2832160e+04],
       [2.0200000e+03, 0.0000000e+00, 1.2454907e+05],
       [2.0060000e+03, 2.5757000e+04, 9.2612100e+04],
       [2.0120000e+03, 1.0728000e+04, 9.7497730e+04],
       [2.0200000e+03, 0.0000000e+00, 5.6445200e+04],
       [2.0090000e+03, 7.7599000e+04, 1.1231044e+05],
       [2.0100000e+03, 9.9197000e+04, 1.2071627e+05],
       [2.0110000e+03, 3.7978000e+04, 7.6566490e+04],
       [2.0020000e+03, 1.2859000e+04, 7.1647590e+04],
       [2.0070000e+03, 8.0520000e+03, 7.3919530e+04],
       [2.0010000e+03, 8.9773000e+04, 1.1273299e+05],
       [2.0200000e+03, 0.0000000e+00, 5.3183380e+04],
       [2.0090000e+03, 4.1457000e+04, 1.2748842e+05],
       [2.0160000e+03, 1.1560700e+05, 5.9910400e+04],
       [2.0120000e+03, 4.6449000e+04, 6.1118590e+04],
       [2.0200000e+03, 0.0000000e+00, 8.8552390e+04],
       [2.0160000e+03, 3.708

In [544]:
dataset.shape

(258, 3)

## *np.mean()*

Devuelve el promedio de los elementos del array a lo largo del eje especificado.

In [545]:
np.mean(dataset)

48489.225297157616

Este valor no tiene sentido porque estamos combinando años kilometraje y valor de los vehículos.

**Con el atributo axis le indicamos que la media sea realizada de forma columnar(vertical) con el valor 0 o horizontal con el valor 1.**

Como tenemos los valores tenemos separados en columnas entonces quiero que calcule la media de cada columna.

In [546]:
np.mean(dataset, axis=0)

array([ 2007.74806202, 44499.41472868, 98960.51310078])

Ahora sí tenemos una media relacionada con la información de los vehículos

Se puede realizar la medio horizontal pero no tienen sentido porque estaría combinando años kilometrajes y valor.

Otra forma de visualizar el dataset es particionándolo.

In [552]:
np.mean(dataset[ : : , 0])

2007.7480620155038

In [547]:
np.mean(dataset[ : : , 1])

44499.41472868217

In [548]:
np.mean(dataset[ : : , 2])

98960.51310077519

## *np.std()*

Devuelve la desviación estándar de los elementos del array a lo largo del eje especificado.

In [550]:
np.std(dataset[ : : , 0])

9.991978183881399

In [551]:
np.std(dataset[ : : , 1])

39859.82699005149

In [549]:
np.std(dataset[ : : , 2])

29754.101150388564

## *ndarray.sum()*

Devuelve la suma de los elementos del array a lo largo del eje especificado.

In [553]:
dataset.sum(axis=0)

array([  517999.        , 11480849.        , 25531812.37999999])

Sumando utilizando particiones.  
dataset.sum(lineas, columnas)

Ahora no va a sumar todo el dataset solamente la columna 1 que le estoy indicando.

In [554]:
dataset[ : :, 1].sum()

11480849.0

## *np.sum()*

Devuelve la suma de los elementos del array a lo largo del eje especificado.

In [555]:
np.sum(dataset, axis=0)

array([  517999.        , 11480849.        , 25531812.37999999])

In [556]:
np.sum(dataset[ : :, 2])

25531812.38

## **Biblioteca Pandas**

In [558]:
import pandas as pd


In [560]:
df = pd.DataFrame(data=dataset, columns=['años', 'km', 'valor'])
result = df.groupby('años').mean()
print(result)

                  km          valor
años                               
1990.0  72083.666667   96116.206667
1991.0  47036.444444   93420.658889
1992.0  51283.000000   90670.667500
1993.0  55384.250000   93357.635000
1994.0  70260.777778   96623.326667
1995.0  71289.714286   86117.702857
1996.0  50926.100000   98619.528000
1997.0  71897.000000   96852.883333
1998.0  71828.000000   97801.913333
1999.0  39075.500000   98675.220000
2000.0  98921.000000  112101.906667
2001.0  69583.750000  105120.757500
2002.0  59268.545455   89327.512727
2003.0  56926.571429  111672.687143
2004.0  62735.800000   82253.320000
2005.0  58052.500000  117836.888333
2006.0  45569.750000   91200.577500
2007.0  59986.750000  101135.170000
2008.0  61647.833333  103123.380000
2009.0  48694.000000  110470.614286
2010.0  79227.333333   98001.035000
2011.0  53031.000000  106613.244000
2012.0  40817.500000   89212.192500
2013.0  39021.777778  105672.001111
2014.0  57461.500000   87228.613333
2015.0  61032.125000  103795